drivers/rtc/rtc_numaker.c

```diff
@@ -0,0 +1,401 @@
+				      const struct rtc_time *timeptr)
+{
+	struct rtc_numaker_data *data = dev->data;
+	const struct rtc_numaker_config *config = dev->config;
+	RTC_T *rtc_base = config->rtc_base;
+	uint16_t mask_capable;
+	struct rtc_numaker_time alarm_time;
+
+	rtc_numaker_alarm_get_supported_fields(dev, 0, &mask_capable);
+
+	if ((id != 0)) {
+		return -EINVAL;
+	}
+
+	if ((mask != 0) && (timeptr == NULL)) {
+		return -EINVAL;
+	}
+
+	if (mask & ~mask_capable) {
+		return -EINVAL;
+	}
+
+	if (rtc_utils_validate_rtc_time(timeptr, mask) == false) {
+		return -EINVAL;
+	}
+
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	if ((mask == 0) || (timeptr == NULL)) {
+		/* Disable the alarm */
+		data->camsk = 0x00;
+		data->tamsk = 0x00;
+		data->alarm_mask = mask;
+		rtc_base->CAMSK = data->camsk;
+		rtc_base->TAMSK = data->tamsk;
+		/* Disable RTC Alarm Interrupt */
+		RTC_DisableInt(RTC_INTEN_ALMIEN_Msk);
+		k_spin_unlock(&data->lock, key);
+		return 0;
+	}
+
+	alarm_time.time_scale = NVT_TIME_SCALE;
+	RTC_GetDateAndTime((S_RTC_TIME_DATA_T *)&alarm_time);
+
+	/* Reset RTC alarm mask */
+	data->camsk = NVT_ALARM_MSK;
+	data->tamsk = NVT_ALARM_MSK;
+
+	/* Set H/W care to match bits */
+	if (mask & RTC_ALARM_TIME_MASK_YEAR) {
+		alarm_time.year = timeptr->tm_year + TM_YEAR_REF;
+		data->camsk &= ~(NVT_ALARM_UNIT_MSK << RTC_CAMSK_MYEAR_Pos);
+	}
+	if (mask & RTC_ALARM_TIME_MASK_MONTH) {
+		alarm_time.month = timeptr->tm_mon + 1;
+		data->camsk &= ~(NVT_ALARM_UNIT_MSK << RTC_CAMSK_MMON_Pos);
+	}
+	if (mask & RTC_ALARM_TIME_MASK_MONTHDAY) {
+		alarm_time.day = timeptr->tm_mday;
+		data->camsk &= ~(NVT_ALARM_UNIT_MSK << RTC_CAMSK_MDAY_Pos);
+	}
+	if (mask & RTC_ALARM_TIME_MASK_HOUR) {
+		alarm_time.hour = timeptr->tm_hour;
+		data->tamsk &= ~(NVT_ALARM_UNIT_MSK << RTC_TAMSK_MHR_Pos);
+	}
+	if (mask & RTC_ALARM_TIME_MASK_MINUTE) {
+		alarm_time.minute = timeptr->tm_min;
+		data->tamsk &= ~(NVT_ALARM_UNIT_MSK << RTC_TAMSK_MMIN_Pos);
+	}
+	if (mask & RTC_ALARM_TIME_MASK_SECOND) {
+		alarm_time.second = timeptr->tm_sec;
+		data->tamsk &= ~(NVT_ALARM_UNIT_MSK << RTC_TAMSK_MSEC_Pos);
+	}
+
+	/* Disable RTC Alarm Interrupt */
+	RTC_DisableInt(RTC_INTEN_ALMIEN_Msk);
+
+	/* Set the alarm time */
+	RTC_SetAlarmDateAndTime((S_RTC_TIME_DATA_T *)&alarm_time);
+
+	/* Clear RTC alarm interrupt flag */
+	RTC_CLEAR_ALARM_INT_FLAG();
+
+	data->alarm_time = *timeptr;
+	data->alarm_mask = mask;
+
+	rtc_base->CAMSK = data->camsk;
+	rtc_base->TAMSK = data->tamsk;
+
+	k_spin_unlock(&data->lock, key);
+
+	/* Enable RTC Alarm Interrupt */
+	RTC_EnableInt(RTC_INTEN_ALMIEN_Msk);
+
+	return 0;
+}
+
+static int rtc_numaker_alarm_get_time(const struct device *dev, uint16_t id, uint16_t *mask,
+				      struct rtc_time *timeptr)
+{
+	struct rtc_numaker_data *data = dev->data;
+
+	if ((id != 0) || (mask == NULL) || (timeptr == NULL)) {
+		return -EINVAL;
+	}
+
+	K_SPINLOCK(&data->lock) {
+		*timeptr = data->alarm_time;
+		*mask = data->alarm_mask;
+	}
```

Comment:
Is it possible to read the alarm configuration from registers? the cached values will not survive a reboot I suspect, which the alarm configuration most likely will :)